apps/builder/app/builder/features/props-panel/use-props-logic.ts

```diff
@@ -1,156 +1,240 @@
+  return value;
 };
 
 /**
- * usePropsLogic expects that key={selectedInstanceData.id} is used on the ancestor component
+ * usePropsLogic expects that key={instance.id} is used on the ancestor component
  */
 export const usePropsLogic = ({
-  props,
-  selectedInstance,
-  updateProps,
+  props: savedProps,
+  instance,
+  updateProp,
   deleteProp,
-}: UsePropsLogic) => {
-  const { id: instanceId, component } = selectedInstance;
+}: UsePropsLogicInput): UsePropsLogicOutput => {
+  const meta = getComponentPropsMeta(instance.component);
+  if (meta === undefined) {
+    throw new Error(`Could not get meta for compoent "${instance.component}"`);
+  }
 
-  const meta = getComponentPropsMeta(component);
-  const metaProps = meta?.props ?? {};
-  const initialProps = meta?.initialProps ?? [];
+  // we will delete items from these maps as we categorize the props
+  const unprocessedSaved = new Map(savedProps.map((prop) => [prop.name, prop]));
+  const unprocessedKnown = new Map(Object.entries(meta.props));
 
-  const initialPropsByName = new Map<string, Prop>();
-  for (const name of initialProps) {
-    const prop = getPropFromMetaProps(instanceId, metaProps, name);
-    if (prop) {
-      initialPropsByName.set(name, prop);
-    }
-  }
+  const initialProps: PropAndMeta[] = (meta.initialProps ?? []).map((name) => {
+    const saved = getAndDelete(unprocessedSaved, name);
+    const known = getAndDelete(unprocessedKnown, name);
 
-  const [addedProps, setAddedProps] = useState<Prop[]>([]);
-  const addedPropsById = new Map<string, Prop>();
-  for (const prop of addedProps) {
-    addedPropsById.set(prop.id, prop);
-  }
+    if (known === undefined) {
+      throw new Error(
+        `The prop "${name}" is defined in meta.initialProps but not in meta.props`
+      );
+    }
 
-  const propsById = new Map<string, Prop>();
-  for (const prop of props) {
-    if (initialPropsByName.has(prop.name)) {
-      initialPropsByName.set(prop.name, prop);
-      continue;
+    let prop = saved;
+
+    // For initial props, if prop is not saved, we want to show default value if available.
+    //
+    // Important to not use getStartingProp if default is not available
+    // beacuse user may have this experience:
+    //   - they open props panel of an Image
+    //   - they see 0 in the control for "width"
+    //   - where 0 is a fallback when no default is available
+    //   - they think that width is set to 0, but it's actually not set at all
+    //
+    if (prop === undefined && known.defaultValue !== undefined) {
+      prop = getStartingProp(instance.id, known, name);
     }
-    if (addedPropsById.has(prop.id)) {
-      addedPropsById.set(prop.id, prop);
-      continue;
+
+    return { prop, propName: name, meta: known };
+  });
+
+  // names of the props added by the user during the lifetime of the hook
+  const [newNames, setNewNames] = useState<Prop["name"][]>([]);
+
+  const oldOptional: PropAndMeta[] = [...unprocessedSaved.values()]
```

Comment:
isn't values() already returning a new array essentially, why clone it?

Reply:
ah sorry, its an iterator convertion to array, maybe better Array.from() to be clear